docs/adr/adr_backend_etl_and_hosting_for_StatViz.md

```diff
@@ -114,124 +116,185 @@ beneficiaryDemographics {
 }
 ```
 
-### B. Data Transformation
+### B. Data Structure and Transformation
+
+Our current method of tracking changes to database records, like stock or people, employs a single, centralized "history" table. Its design is as follows:
+
+```
+CREATE TABLE `history` (
+  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
+  `tablename` varchar(255) DEFAULT NULL,
+  `record_id` int(11) DEFAULT NULL,
+  `changes` text,
+  `user_id` int(11) unsigned DEFAULT NULL,
+  `ip` varchar(20) DEFAULT NULL,
+  `changedate` datetime DEFAULT NULL,
+  `from_int` int(11) DEFAULT NULL,
+  `to_int` int(11) DEFAULT NULL,
+  `from_float` float DEFAULT NULL,
+  `to_float` float DEFAULT NULL,
+)
+```
+
+For instance, when a box transitions from one warehouse to an external location, the "history" table captures it as:
+
+```
+(..., tablename, record_id, changes, ..., changedate, from_int, to_int, ...)
+(..., "stock", "Box1", "location_id", ..., "2023-08-07 12:00:00", 1, 2) // change from WH1 to Free Shop
+(..., "stock", "Box1", "box_state_id", ..., "2023-08-07 12:00:00", 1, 2) // change from InStock to Donated
+```
+
+This data structure efficiently captures transactional data changes, but extracting historical snapshots (like warehouse stock at a particular time) is more complex.
+
+In order to figure out whether the current "history" table is "enough", we started with directly transforming operational data from Boxtribute in response to incoming graphQL requests. We did not cache any for these transformations or create additional data-structure.
+
+We learned (see PR #1013) that
+
+- the structure of the "history" table already allows us to reproduce old states of data
 
-The BE has to transform Boxtribute's data to be able to return the data in the desired fromat from above. Now, the question arises when and how to do these transformations. We see the following options:
+  --> the data structure is compatible to respond to (probably) all queries.
 
-#### B.1 On-the-fly Transformation During GraphQL Requests
+- However, it comes at a great cost (see PR #1013). Even with a small sample size the latency of recreating old states of the data with the transactional data from the "history" table lead to an unacceptable latency of >10s.
 
-This is straight-forward solution and how we started for the first visualizations - Directly transform operational data from Boxtribute in response to incoming graphQL requests.
+  -->**We need to add a cache or extend the schema** to make the analytical queries run faster.
+
+### B.1 Extending the DB schema
+
+Especially querying historical snapshots of data is read-intensive with the transactional "history" table.
+Therefore, to reduce the latency of such queries we consider adding data structures to the schema that make querying snapshots easier.
+
+#### B.1.1 Adding Bi-temporal / Shadow Tables
```

Comment:
After reading [this blog post](https://martinfowler.com/articles/bitemporal-history.html) I don't think bitemporal tables solve the issue we have, I think we want plain shadow/temporal tables.

Reply:
Hey @pylipp I read the link and I believe what David means with bi-temporal table is not the same as in the blog post. Bi-temporal means that it saves two dates. For David the two dates are "from" and "to" datetime so that we only have to compare columns and not rows. In the blog post it is about record date and actual date which I agree is not a concept that helps us.